server/auth/routes.py

```diff
@@ -188,24 +182,87 @@ def auditadmin_login_callback():
     return redirect("/")
 
 
-@auth.route("/auth/jurisdictionadmin/start")
-def jurisdictionadmin_login():
-    redirect_uri = urljoin(request.host_url, JURISDICTIONADMIN_OAUTH_CALLBACK_URL)
-    return auth0_ja.authorize_redirect(redirect_uri=redirect_uri)
+def generate_login_code(timestamp: datetime):
+    return HOTP(LOGIN_CODE_SECRET).at(
+        int(timestamp.timestamp() / LOGIN_CODE_LIFETIME.total_seconds())
+    )
 
 
-@auth.route(JURISDICTIONADMIN_OAUTH_CALLBACK_URL)
-def jurisdictionadmin_login_callback():
-    auth0_ja.authorize_access_token()
-    resp = auth0_ja.get("userinfo")
-    userinfo = resp.json()
+def verify_login_code(timestamp: datetime, code: str):
+    return HOTP(LOGIN_CODE_SECRET).verify(
+        code, int(timestamp.timestamp() / LOGIN_CODE_LIFETIME.total_seconds())
+    )
 
-    if userinfo and userinfo["email"]:
-        user = User.query.filter_by(email=userinfo["email"]).first()
-        if user and len(user.jurisdiction_administrations) > 0:
-            set_loggedin_user(session, UserType.JURISDICTION_ADMIN, userinfo["email"])
 
-    return redirect("/")
+@auth.route("/auth/jurisdictionadmin/code", methods=["POST"])
+def jurisdiction_admin_generate_code():
+    body = request.get_json()
+    user = User.query.filter_by(email=body.get("email")).one_or_none()
+    if user is None:
+        raise BadRequest(
+            "This email address is not authorized to access Arlo."
+            " Please check that you typed the email correctly,"
+            " or contact your Arlo administrator for access."
+        )
+
+    if user.login_code_requested_at is None or (
+        # Reuse the existing login code if it hasn't expired yet. That way if
+        # they request a new code while waiting for a slow email, we won't wipe
+        # out the code we sent when the email does come through.
+        datetime.now(timezone.utc) - user.login_code_requested_at
+        > LOGIN_CODE_LIFETIME
+    ):
+        user.login_code_requested_at = datetime.now(timezone.utc)
+
+    user.login_code_attempts = 0
+
+    code = generate_login_code(user.login_code_requested_at)
+    email_response = requests.post(
+        f"https://api.mailgun.net/v3/{MAILGUN_DOMAIN}/messages",
+        auth=("api", MAILGUN_API_KEY),
+        data={
+            "from": "Arlo Support <rla@vx.support>",
+            "to": [user.email],
+            "subject": "Welcome to Arlo - Use the Code in this Email to Log In",
+            "text": render_template("email_login_code.txt", code=code),
+            "html": render_template("email_login_code.html", code=code),
+        },
+    )
+    email_response.raise_for_status()
+
+    db_session.commit()
+
+    return jsonify(status="ok")
+
+
+@auth.route("/auth/jurisdictionadmin/login", methods=["POST"])
+def jurisdiction_admin_login():
+    body = request.get_json()
+    user = User.query.filter_by(email=body.get("email")).one_or_none()
```

Comment:
There's a race condition here that could be security problematic: if you don't lock the row, then multiple parallel requests could trounce each other and allow an attacker to brute force more than the limit of tries in parallel.

Can we do a select for update on the `user`?

Reply:
yep, will do